src/Symfony/Component/DependencyInjection/Compiler/MergeExtensionConfigurationPass.php

```diff
@@ -92,12 +92,16 @@ public function __construct(parent $parameterBag)
         $this->mergeEnvPlaceholders($parameterBag);
     }
 
-    public function freezeAfterProcessing(Extension $extension)
+    public function freezeAfterProcessing(Extension $extension, ContainerBuilder $container)
     {
+        if (!$config = $extension->getProcessedConfigs()) {
+            // Extension::processConfiguration() wasn't called, we cannot know how configs were merged
+            return;
+        }
         $this->processedEnvPlaceholders = array();
 
-        // serialize config to catch env vars nested in object graphs
-        $config = serialize($extension->getProcessedConfigs());
+        // serialize config and container to catch env vars nested in object graphs
+        $config = serialize($config).serialize($container->getDefinitions()).serialize($container->getAliases()).serialize($container->getParameterBag()->all());
```

Comment:
aliases don't need to be serialized. They only have a target and a public flag, and these cannot use parameters.

Reply:
They can somehow, see https://github.com/nicolas-grekas/symfony/blob/a8e6aac2f58ca3f12c931f20a6de29ef46686dc8/src/Symfony/Component/DependencyInjection/Compiler/ResolveParameterPlaceHoldersPass.php